scylla/src/transport/connection.rs

```diff
@@ -820,6 +822,59 @@ impl Connection {
         }
     }
 
+    async fn prepare_batch<'b>(
+        &self,
+        init_batch: &'b Batch,
+        values: impl BatchValues,
+    ) -> Result<Cow<'b, Batch>, QueryError> {
+        let mut to_prepare = HashSet::<String>::new();
+
+        {
+            let mut values_iter = values.batch_values_iter();
+            for stmt in &init_batch.statements {
+                if let BatchStatement::Query(query) = stmt {
+                    let value = values_iter.next_serialized().transpose()?;
+                    if let Some(v) = value {
+                        if v.len() > 0 {
+                            to_prepare.insert(query.contents.clone());
+                        }
+                    }
+                } else {
+                    values_iter.skip_next();
+                }
+            }
+        }
+
```

Comment:
How about an early return here?

```rust
if to_prepare.is_empty() {
    return Cow::Borrowed(init_batch);
}
```

By doing that we can skip creating `prepared_queries` and the loop that follows after it altogether on the fast path (i.e. when all statements are prepared).